src/transport.rs

```
@@ -123,6 +123,37 @@ impl KKTransport {
     pub fn remote_static(&self) -> NoisePubKey {
         self.channel.remote_static()
     }
+
+    /// Read a message from the other end of the encrypted communication channel.
+    /// Will recover from certain types of errors, those for which no bytes are
+    /// read from the stream, by retrying up to 10 times with a 1s sleep between
+    /// attempts. After 10 attempts, or an unrecoverable error, will return an
+    /// error.  
+    pub fn read(&mut self) -> Result<Vec<u8>, Error> {
+        for _attempt in 0..9 {
+            match self._read() {
+                Ok(msg) => return Ok(msg),
+                Err(error) => match error {
+                    Error::Noise(e) => return Err(Error::Noise(e)),
+                    Error::Snow(e) => return Err(Error::Snow(e)),
+                    Error::Message(e) => return Err(Error::Message(e)),
+                    Error::Transport(e) => match e.kind() {
+                        ErrorKind::UnexpectedEof => return Err(Error::Transport(e)),
+                        ErrorKind::Interrupted => return Err(Error::Transport(e)),
+                        _ => {
+                            thread::sleep(Duration::from_secs(1));
+                            continue;
+                        }
+                    },
+                },
+            };
+        }
+
+        Err(Error::Transport(std::io::Error::new(
+            ErrorKind::Other,
+            "Abort read after too many failed attempts",
+        )))
```

Just use io::Error::TimedOut ? But actually i think we should return the last error, it's more useful to the caller.